webrender/src/device/gfx/device.rs

```diff
@@ -1492,90 +1453,105 @@ impl<B: hal::Backend> Device<B> {
-                    ) {
-                        unsafe {
-                            cmd_buffer.pipeline_barrier(
-                                src_stage.unwrap()
-                                    .. hal::pso::PipelineStage::FRAGMENT_SHADER,
-                                hal::memory::Dependencies::empty(),
-                                &[barrier],
-                            );
-                        }
-                    }
-                }
-                program.bound_textures[index] = self.bound_textures[index];
-            }
-            need_alloc
-        };
-        if need_alloc {
-            self.descriptor_pools_per_draw[self.next_id]
-                .next(shader_group, &self.device, &self.pipeline_requirements);
-        }
-
-        let (desc_set, _) = self.descriptor_pools_per_frame[self.next_id].get_set_by_group(shader_group);
-        for &(index, sampler_name) in SAMPLERS[PER_DRAW_SAMPLER_COUNT..].iter() {
-            program.bind_texture(
-                &self.device,
-                desc_set,
-                &self.images[&self.bound_textures[index]].core,
-                sampler_name,
+            self.per_pass_descriptors.bind_textures(
+                &self.bound_textures,
                 &mut cmd_buffer,
+                per_pass_bindings,
+                &self.images,
+                &mut self.desc_allocator,
+                &self.device,
+                &self.desc_group_data,
+                &descriptor_group,
+                DESCRIPTOR_SET_PER_PASS,
+                PER_DRAW_TEXTURE_COUNT..PER_DRAW_TEXTURE_COUNT + PER_PASS_TEXTURE_COUNT,
+                Some(&self.sampler_linear),
             );
-            program.bound_textures[index] = self.bound_textures[index];
-        }
+            self.bound_per_pass_textures = per_pass_bindings;
+        }
+
+        // Per frame textures
+        let per_group_bindings = PerGroupBindings(
+            [
+                self.bound_textures[5],
+                self.bound_textures[6],
+                self.bound_textures[7],
+                self.bound_textures[8],
+                self.bound_textures[9],
+                self.bound_textures[10],
+            ],
+        );
+
+        self.per_group_descriptors.bind_textures(
+            &self.bound_textures,
+            &mut cmd_buffer,
+            (descriptor_group, per_group_bindings),
+            &self.images,
+            &mut self.desc_allocator,
+            &self.device,
+            &self.desc_group_data,
+            &descriptor_group,
+            DESCRIPTOR_SET_PER_GROUP,
+            match descriptor_group {
+                DescriptorGroup::Default => PER_GROUP_RANGE_DEFAULT,
+                DescriptorGroup::Clip => PER_GROUP_RANGE_CLIP,
+                DescriptorGroup::Primitive => PER_GROUP_RANGE_PRIMITIVE,
+            },
+            Some(&self.sampler_nearest),
+        );
+        self.bound_per_group_textures = per_group_bindings;
+
+        // Immutable samplers
```

Comment:
immutable samplers should be a part of the descriptor set layout, never need to be bound specifically
or perhaps these is about some other form of immutability?

Reply:
No, it's a typo, this should be `// Mutable samplers`.